agent/engine/host_resource_manager.go

```diff
@@ -16,20 +16,226 @@
 package engine
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/aws/amazon-ecs-agent/agent/ecs_client/model/ecs"
 	"github.com/aws/amazon-ecs-agent/agent/logger"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 )
 
+// TODO remove this once resource, consume are used
+//lint:file-ignore U1000 Ignore all unused code
+
 // initialHostResource keeps account of each task in
 type HostResourceManager struct {
-	initialHostResource map[string]*ecs.Resource
-	consumedResource    map[string]*ecs.Resource
+	initialHostResource       map[string]*ecs.Resource
+	consumedResource          map[string]*ecs.Resource
+	hostResourceManagerRWLock sync.Mutex
 
 	//task.arn to boolean whether host resources consumed or not
 	taskConsumed map[string]bool
 }
 
+// Returns if resources consumed or not and error status
+// false, nil -> did not consume, should stay pending
+// false, err -> resources map has errors, something went wrong
+// true, nil -> successfully consumed
+func (h *HostResourceManager) consume(taskArn string, resources map[string]*ecs.Resource) (bool, error) {
+	h.hostResourceManagerRWLock.Lock()
+	defer h.hostResourceManagerRWLock.Unlock()
+
+	defer logger.Debug("Consumed resources after task consume call", logger.Fields{
+		"taskArn":   taskArn,
+		"CPU":       *h.consumedResource["CPU"].IntegerValue,
+		"MEMORY":    *h.consumedResource["MEMORY"].IntegerValue,
+		"PORTS_TCP": h.consumedResource["PORTS_TCP"].StringSetValue,
+		"PORTS_UDP": h.consumedResource["PORTS_UDP"].StringSetValue,
+		"GPU":       *h.consumedResource["GPU"].IntegerValue,
+	})
+
+	// Check if already consumed
+	_, ok := h.consumedResource[taskArn]
+	if ok {
+		// Nothing to do, already consumed, return
+		return true, nil
+	}
+
+	ok, err := h.consumable(resources)
+	if err != nil {
+		return false, err
+	}
+	if ok {
+		// CPU
+		cpu := *h.consumedResource["CPU"].IntegerValue + *resources["CPU"].IntegerValue
+		h.consumedResource["CPU"].SetIntegerValue(cpu)
+
+		// MEM
+		mem := *h.consumedResource["MEMORY"].IntegerValue + *resources["MEMORY"].IntegerValue
+		h.consumedResource["MEMORY"].SetIntegerValue(mem)
+
+		// PORTS
+		portsResource, ok := resources["PORTS_TCP"]
```

Comment:
How about defining constants for resource identifiers like `"PORTS_TCP"`? Using the identifiers directly seems difficult to maintain and error-prone.